internal/rundata/dut.go

```diff
@@ -0,0 +1,251 @@
+// OpenConfig paths.
+type dutInfo struct {
+	vendor string
+	model  string
+	osver  string
+}
+
+// setFromComponentChassis sets dutInfo from the first component of type CHASSIS.
+//
+//   - vendor from mfg-name (Arista, Cisco).
+//   - model from either description (Cisco, Juniper) or part-no (Arista).
+//   - osver from software-version (Juniper).
+func (di *dutInfo) setFromComponentChassis(ctx context.Context, y components.Y) {
+	if di.vendor != "" && di.model != "" && di.osver != "" {
+		return // No-op if nothing needs to be set.
+	}
+
+	const chassisType = oc.PlatformTypes_OPENCONFIG_HARDWARE_COMPONENT_CHASSIS
+
+	chassisIDs, err := y.FindByType(ctx, chassisType)
+	if err != nil {
+		glog.Errorf("Could not find chassis: %v", err)
+		return
+	}
+
+	chassisID := chassisIDs[0]
+
+	componentPath := ocpath.Root().Component(chassisID)
+	component, err := ygnmi.Get(ctx, y.Client, componentPath.State())
+	if err != nil {
+		glog.Errorf("Could not get chassis: %v", err)
+		return
+	}
+
+	if glog.V(2) {
+		if out, err := json.MarshalIndent(component, "", "  "); err == nil {
+			glog.Info(string(out))
+		}
+	}
+
+	if di.vendor == "" && component.MfgName != nil {
+		di.vendor = *component.MfgName
+		glog.V(2).Infof("Setting vendor from chassis mfg-name: %s", di.vendor)
+	}
+
+	if di.model == "" && component.PartNo != nil {
+		di.model = *component.PartNo
+		glog.V(2).Infof("Setting model from chassis part-no: %s", di.model)
+	}
+
+	if desc := component.Description; desc != nil {
+		if di.model == "" || strings.HasPrefix(*desc, "Cisco ") {
+			di.model = *component.Description
+			glog.V(2).Infof("Setting model from chassis description: %s", di.model)
+		}
+	}
+
+	if di.osver == "" && component.SoftwareVersion != nil {
+		di.osver = *component.SoftwareVersion
+		glog.V(2).Infof("Setting osver from chassis software-version: %s", di.osver)
+	}
+}
+
+// setFromComponentOS sets dutInfo from the first component of type OPERATING_SYSTEM, osver from
+// software-version (Arista, Cisco).
+func (di *dutInfo) setFromComponentOS(ctx context.Context, y components.Y) {
+	if di.osver != "" {
+		return // No-op if osver is already set.
+	}
+
+	const osType = oc.PlatformTypes_OPENCONFIG_SOFTWARE_COMPONENT_OPERATING_SYSTEM
+
+	osIDs, err := y.FindByType(ctx, osType)
+	if err != nil {
+		glog.Errorf("Could not find operating-system: %v", err)
+		return
+	}
+
+	osID := osIDs[0]
+
+	softVerPath := ocpath.Root().Component(osID).SoftwareVersion()
+	softVer, err := ygnmi.Get(ctx, y.Client, softVerPath.State())
+	if err != nil {
+		glog.Errorf("Missing component software-version: %v", err)
+	} else {
+		di.osver = softVer
+		glog.V(2).Infof("Setting osver from operating-system software-version: %s", di.osver)
+	}
+}
+
+// setFromLLDP sets dutInfo vendor from LLDP system-description (Juniper).  This is less
+// reliable because LLDP config can be changed.
+func (di *dutInfo) setFromLLDP(ctx context.Context, y components.Y) {
```

Comment:
Should we add `Juniper` in the function name as well? The function name looks like a general use function but logic seems very specific to Juniper.

Reply:
It's supposed to be a general use function. There could be another vendor that needs the LLDP fallback.